controlplane/kubeadm/internal/etcd_client_generator.go

```diff
@@ -92,45 +92,57 @@ func (c *EtcdClientGenerator) forLeader(ctx context.Context, nodeNames []string)
 	// Loop through the existing control plane nodes.
 	var errs []error
 	for _, nodeName := range nodeNames {
-		// Get a temporary client to the etcd instance hosted on the node.
-		client, err := c.forFirstAvailableNode(ctx, []string{nodeName})
+		cl, tryNext, err := c.tryGetLeaderClient(ctx, nodeName, nodes)
 		if err != nil {
-			errs = append(errs, err)
-			continue
+			if tryNext {
+				errs = append(errs, err)
+				continue
+			}
+			return nil, err
 		}
-		defer client.Close()
 
-		// Get the list of members.
-		members, err := client.Members(ctx)
-		if err != nil {
-			errs = append(errs, err)
-			continue
-		}
+		return cl, nil
+	}
+	return nil, errors.Wrap(kerrors.NewAggregate(errs), "could not establish a connection to the etcd leader")
+}
 
-		// Get the leader member.
-		var leaderMember *etcd.Member
-		for _, member := range members {
-			if member.ID == client.LeaderID {
-				leaderMember = member
-				break
-			}
-		}
+func (c *EtcdClientGenerator) tryGetLeaderClient(ctx context.Context, nodeName string, allNodes sets.String) (client *etcd.Client, tryNext bool, err error) {
```

Comment:
Instead of having 3 return values, have we considered to either close the client every time we continue (in addition to the defer) or use a specialized error wrapper that we can check with `errors.Is()`?

Reply:
I briefly considered closing the client explicitly at all locations where the loop continues or the method terminates but found that relatively brittle and hard to maintain (especially for future maintainers that may extend the logic).

I like the idea to wrap the error, that seems more elegant than having three return values. I pushed another commit that implements your idea (with the small delta that I used `errors.As` so that we can continue to aggregate the error messages).

Please take another look.